woocommerce/woocommerce
# Update shipping and payment radio controls to use borders on selected items

### Changes proposed in this Pull Request:

In this PR I propose some changes to the radio control and radio accordion (CC @samueljseay as you may be looking at these as part of the `@wordpress/components` project)

The main changes are:
- Add a new prop `RadioControl` and `RadioControlAccordion` called `highlightChecked` which defaults to false.
  - This is to ensure we can explicitly enable this setting on components we choose, rather than blanket applying it to all instances.
- Update components that use these and want to highlight the selected elements to pass this prop

The rest of the changes apply only when `highlightChecked` is true:

- Enclose the entire radio group in a fake border implemented with the `::after` pseudo element. 
  - This is because using regular CSS borders or the `outline` property resulted in double borders on the selected element. Using a pseudo element allows us finer control over the border compared to what we'd get with the border property.
- Add an `::after` element to each item in the group to replace the `border-top` property. This is to add a separator between each item. Using the border property resulted in a small area where two transparent colours would combine to make a darker colour. The pseudo element gives us better control and allows us to bring the "border" in by 2px to prevent this.
- Add new class names to the radio control group to signal when the first or last element is selected. This is so we can remove the top or bottom border to prevent double borders.
- Use box-shadow to add a highlight to the selected element. This uses `currentColor` so it follows the theme.





Closes #46009.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure you have multiple shipping methods set up. (Go to WooCommerce -> Settings -> Shipping and add several methods to a zone, you can create several flat rates. Give them different names and prices).
2. Enable Local pickup (Go to WooCommerce -> Settings -> Shipping -> Local Pickup). Add 3 or more locations. Enable them all.
3. Ensure you have several payment methods set up (e.g. Stripe, Bank Transfer, COD etc.) Ensure one of them supports saved payment methods (e.g. Stripe)
4. Check out three times using the payment method that supports saved options. Each time, use a different carf so it can save three different cards. [For reference, here are some cards you can use with Stripe](https://docs.stripe.com/testing#cards).
5. Add an item to your cart and go to the Cart block.
6. Ensure the shipping rates on the sidebar show correctly. They should not have borders and should look like the bottom right of the image.

<img width="885" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/9badba85-c8d7-456a-98e4-0f1e646c6537">


7. Try changing the shipping methods in the cart. It should work without any errors. If you see a notice about local pickup, that's OK.
8. Proceed to the Checkout block.
9. Select "Shipping" for your shipping method.
10. Scroll down to the shipping methods and change it a few times. Ensure they display like the screenshot below. There should be no border overlaps or double borders. 

<img width="559" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/e778da39-3295-4186-a377-6e537b318c8d">

11. Change your shipping method to "Local pickup" and repeat the test with the pickup locations. They should look like this:

<img width="557" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/80651993-d873-4c3a-8e7e-012c9f807342">

12. Change between them and ensure there are no overlapping borders, jumping text etc.
13. Scroll down to the payment method section. Ensure it looks like the screenshots below:

<img width="585" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/8aa3add1-43d0-4036-9d23-f843f79d6530">
and 
<img width="578" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/99efc3e5-bc8b-4de2-8c6a-7e4f5e1b0533">

14. Change the methods a few times, between saved methods and "other" methods. Ensure there are no overlapping borders. The text becoming bold when the option is selected is expected.
15. Install [Multiple Packages for WooCommerce](https://wordpress.org/plugins/multiple-packages-for-woocommerce/). 
16. Go to WooCommerce -> Settings -> Multiple Packages and enable it. Set "Group by" to "Product (Individual).
17. Add 3 items to your cart and repeat the steps relating to checking shipping methods. Each item should be in its own package and each package should have the correct styling.

#### Testing with WC Subscriptions (Don't do this until https://github.com/Automattic/woocommerce-subscriptions-core/pull/593 is merged

0. Install WC Subscriptions
1. Create 3 subscription items with different lengths, (weekly, yearly, monthly)
2. Add them to your cart and go to the Cart block.
3. Ensure the shipping options in the sidebar show correctly, i.e. without borders around each rate.

<img width="300" alt="image" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/5656702/114f92c1-e584-4b1a-b409-442f59986b5e">

4. Go to the Checkout block. Ensure you can see the additional packages, and that they _do_ have borders around each rate

<img width="517" alt="image" src="https://github.com/Automattic/woocommerce-subscriptions-core/assets/5656702/8c8a0aab-c4df-4f7f-890b-9d666f54eda4">

5. Switch back to WC Core trunk and run the tests again, the borders won't show but please ensure the shipping selectors for recurring packages still work and can change the method.

#### Developer only steps for when WC Subscriptions 593 is not merged

19. Check out https://github.com/Automattic/woocommerce-subscriptions-core/pull/593 in WooCommerce Subscriptions Core.
20. Follow the testing steps from that PR.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
